torch/_functorch/_aot_autograd/traced_function_transforms.py

```diff
@@ -230,6 +230,12 @@ def inner_fn(primals: List[Any], tangents: List[Any]):
         backward_out: Tuple[Tensor, ...] = ()
         # Call the backwards pass
         if grad_primals:
+            functional_tensor_mode = torch.utils._python_dispatch._detect_infra_mode(
+                torch._C._TorchDispatchModeKey.FUNCTIONAL
+            )
+            if functional_tensor_mode is not None:
+                functional_tensor_mode.on_forward_tracing_end()
```

Comment:
I think I was expecting that you would actually have to tweak the outputs of `joint_fn()` a bit. Right now, we pass in tokens as extra tensor inputs to `joint_fn`, and it returns more tokens as extra outputs of the joint. But if we want to signal to the partitioner that some output tokens should be "forward graph outputs" vs "backward graph outputs", I think we might need to do something like this?
```
        return outs + fw_token_outs, [
            next(backward_out_iter) if i else None for i in inputs_needs_grads
        ] + bw_token_outs
```